sisl/physics/state.py

```diff
@@ -305,6 +305,38 @@ def sub(self, idx):
         sub.info = self.info
         return sub
 
+    @property
+    def state_sc(self):
+        r""" Calculate the state coefficients in the auxiliary supercell
+
+        The format of this state vector is usable for `inner` as the ``ket`` argument.
+        The order of the state vectors is equivalent to the auxiliary supercell order.
+
+        See Also
+        --------
+        inner : these state vectors may be used in the inner product
+        tile : equivalent form along a single given axis
+        translate : a single translation of the state vectors
+
+        Examples
+        --------
+        The returned state vectors can be formulated like this:
+
+        >>> state = State(...)
+        >>> state_sc = np.stack([state.translate(isc)
+        ...                      for _, isc in state.parent.sc])
+        >>> assert state_sc.shape == (len(state) * state.parent.n_s, state.shape[1])
+        """
+        k = _a.asarrayd(self.info.get("k", [0]*3))
+        sc = self.parent.sc
+        if np.allclose(k, 0):
+            def translate(isc):
+                return self.state
+        else:
+            def translate(isc):
+                return self.state * exp(2j*_pi * k @ isc)
+        return np.concatenate([translate(isc) for _, isc in sc], axis=0)
```

Comment:
Shouldn't it be `axis=1` here? Such that `self.state_sc.shape == (len(state), len(state) * state.parent.n_s)`?

The first index should naturally be the band index as for `self.state`. The number of bands dnoes not change by extending to the supercell. The second index the orbital index which grows.

Reply:
No, that would leave the ordering wrong. You would have to *increase* the number of bands to `nbands * n_s` to get this correct. Alternatively it should return `(n_s, nbands, ncoeffs)`?

Reply:
or `(nbands, n_s, ncoeffs)`

Reply:
actually, I am here more in favour of `(n_s, nbands, ncoeffs)`

Reply:
I mean you want to do `es.inner(es.state_sc)`, right, then you would have to increase the number of bands, and not number of coefficients, no?

Reply:
I do not understand - why does the number of bands change? I thought we were just representing the *same* states in a larger region of basis orbitals.

Reply:
Agreed, we are representing the same states in a larger region. But you said you wanted to be able to do `es.inner(es.state_sc)`, the kets represents a band, so if you add more coefficients, it won't work, since the matrix dimensions does not fit.  
Your point is why I think this method was not so obvious to use in the first place.

It becomes unnatural since doing `axis=1` would mean the ordering is wrong, `(nbands, ncoeffs * n_s).T == (ncoeffs * n_s, nbands)` would not enable MM with a square `nbands` matrix, say `H.Hk(...)`.

Reply:
My understanding is this:

We have `es.state[n, i]` where `n` is a band index and `i` an orbital in the primitive cell. And `H.tocsr()` has indices `[i, J]` with `i` an orbital index in the primitive cell and `J` one in the supercell.

To compute general matrix elements we need `<s1| A |s2> = s1.inner(matrix=A.tocsr, ket=s2.state_sc)` to compute general matrix elements between the Bloch states `s1, s2`.

What I have in mind is thus that `s2.state_sc` should have indices `[m, J]` such that the general inner product `<s1(n)| A |s2(m)>` takes the form `[n, i] . [i,J] . ([m, J].T)`.

Reply:
Yes ok, so what you want is to use a non-square matrix as the operator. However, I don't really follow what you then get, since `[i, J] . [m, J].T -> [i, m]` and becomes a square matrix again, and you loose the supercell information? No?

Note: generally you shouldn't use `tocsr` for this, rather use `H.Hk(format='sc:csr')`, this will ensure to be working for NC and SOC as well.

Reply:
It seems to me that this is a greater problem than what this PR proposes to do, I think I should delete the `state_sc`, and then another PR can take care of that, then the discussion can continue there, and more ftitingly so I think. :)

Reply:
What I want is basically the matrix elements between primitive cell band indices `n, m` but where the operator `A` is given in "real space". I do not want any phases applied to `A`.

As far as I see this is also the correct approach to do the inner product between states at different k-points with non-ortho basis, i.e.,
```
<k1(n) | k2(m)> = k1.state[n, i].S[i, J].((k2.state_sc[m, J]).T)
```

Reply:
> becomes a square matrix again, and you loose the supercell information? No?

That's the point, I do not want to retain information about the supercell.

Reply:
lets continue discussion in #356